enhancements/builds/disable-jenkins-pipeline.md

```diff
@@ -0,0 +1,249 @@
+enhancement:
+- What changes (in invocations, configurations, API use, etc.) is an existing
+  cluster required to make on upgrade in order to keep previous behavior?
+- What changes (in invocations, configurations, API use, etc.) is an existing
+  cluster required to make on upgrade in order to make use of the enhancement?
+
+Upgrade expectations:
+- Each component should remain available for user requests and
+  workloads during upgrades. Ensure the components leverage best practices in handling [voluntary disruption](https://kubernetes.io/docs/concepts/workloads/pods/disruptions/). Any exception to this should be
+  identified and discussed here.
+- Micro version upgrades - users should be able to skip forward versions within a
+  minor release stream without being required to pass through intermediate
+  versions - i.e. `x.y.N->x.y.N+2` should work without requiring `x.y.N->x.y.N+1`
+  as an intermediate step.
+- Minor version upgrades - you only need to support `x.N->x.N+1` upgrade
+  steps. So, for example, it is acceptable to require a user running 4.3 to
+  upgrade to 4.5 with a `4.3->4.4` step followed by a `4.4->4.5` step.
+- While an upgrade is in progress, new component versions should
+  continue to operate correctly in concert with older component
+  versions (aka "version skew"). For example, if a node is down, and
+  an operator is rolling out a daemonset, the old and new daemonset
+  pods must continue to work correctly even while the cluster remains
+  in this partially upgraded state for some time.
+
+Downgrade expectations:
+- If an `N->N+1` upgrade fails mid-way through, or if the `N+1` cluster is
+  misbehaving, it should be possible for the user to rollback to `N`. It is
+  acceptable to require some documented manual steps in order to fully restore
+  the downgraded cluster to its previous state. Examples of acceptable steps
+  include:
+  - Deleting any CVO-managed resources added by the new version. The
+    CVO does not currently delete resources that no longer exist in
+    the target version.
+
+### Version Skew Strategy
+
+How will the component handle version skew with other components?
+What are the guarantees? Make sure this is in the test plan.
+
+Consider the following in developing a version skew strategy for this
+enhancement:
+- During an upgrade, we will always have skew among components, how will this impact your work?
+- Does this enhancement involve coordinating behavior in the control plane and
+  in the kubelet? How does an n-2 kubelet without this feature available behave
+  when this feature is used?
+- Will any other components on the node change? For example, changes to CSI, CRI
+  or CNI may require updating that component before the kubelet.
+
+## Implementation History
+
+2020-04-23: Provisional enhancement draft
+
+## Drawbacks
+
+The idea is to find the best form of an argument why this enhancement should _not_ be implemented.
+
```

Comment:
Customers who use Jenkins on OpenShift heavily rely on the pipeline strategy in Build
Configs to create their Jenkins jobs. The pipeline strategy has been deprecated for a while to signal customers that they have to move to an alternative approach for declaratively creating Jenkins jobs based on their Jenkinsfiles. Strong pushback and frequent escalations from customers should be expected if this feature is removed. 

Jenkins Operator initiative has been started to help customers with the transition when the pipeline strategy gets removed. However the Jenkins Operator on OpenShift is still in Tech Preview and customers cannot rely on it as an alternative.

Reply:
True @siamaksade , the pipeline strategy is a convenient way to trigger the job creation in Jenkins and a way to provide customers a command line way to start the Jenkins build using `oc` . That's what the sync plugin does for this part.

It has probably not been explained explicitly what is the alternative with the deprecation. Here are the steps actually:
- User needs to create in Jenkins a new pipeline Job, with "Pipeline script definition from SCM" and set the correct git URL.
- User need to clic the "Start build" button in Jenkins console every time he wants to start a build.
- Kubernetes plugin takes care of the rest (starting agent, building, etc...)


It means that:
- `oc new-build` with pipeline strategy will say `Not supported`, it will not create a `BuildConfig` and `oc start-build` will not operate, as there is not `BuildConfig` of that type.
- The console will not reflect the pipeline state, as the `BuildConfig` does not exist.
- User who wants to follow the pipeline progress and status needs to reach Jenkins console.
- For clusters, with existing `BuildConfig` of pipeline strategy, `oc start-build` needs to inform customer that it is not supported anymore and that user has to start build manually from Jenkins console or use [jenkins-cli](https://www.jenkins.io/doc/book/managing/cli/ ) for instance.
- Builds started from Jenkins will still use agents as the agents lifecycle is managed by the Kubernetes plugin that we embed and pre-configure in our Jenkins image.

Talking about the Jenkins Operator to cover this area, the only added value would be for the Jenkins Operator to watch `BuildConfig` and `Build` objects for which strategy is `Pipeline` and then trigger the Jenkins build, which is actually, what we want to deprecate. And, having this feature has not been discussed until now for the Jenkins Operator as it was supposed to be a `kubernetes` compliant operator.

Reply:
@akram the process you described is not declarative. It cannot be stored in a Git repository which is what customers are doing with the BuildConfig pipeline strategy and recreating all those jobs on a new cluster with a single `oc create -f jobs` command. 

This is the main point that impact customers heavily. They simply do not have a declarative way to create Jenkins pipelines anymore and have to create them one by one through the Jenkins UI every single time.

Reply:
What about using the `jenkins-cli` as suggested. It gets `xml` as an input containing the declaration of the job. That's not `oc` for sure, but, it seems to be a reasonnable tradeoff for the price of the deprecation of `Pipeline` build strategy.

Reply:
Proposal has been updated so that we aren't disabling the strategy. Instead we are introducing a feature gate/feature set.